src/memory.rs

```diff
@@ -114,3 +121,51 @@ pub fn alloc_cons<'a, 'b, A>(
         OCaml::new(cr, ocaml_cons)
     }
 }
+
+const BOX_OPS_DYN_DROP: custom_operations = custom_operations {
+    identifier: "_rust_box_dyn_drop\0".as_ptr() as *const i8,
+    finalize: Some(drop_box_dyn),
+    compare: None,
+    hash: None,
+    serialize: None,
+    deserialize: None,
+    compare_ext: None,
+    fixed_length: ptr::null(),
+};
+
+extern "C" fn drop_box_dyn(oval: RawOCaml) {
+    unsafe {
+        let box_ptr = ocaml_sys::field(oval, 1) as *mut Pin<Box<dyn Any>>;
+        ptr::drop_in_place(box_ptr);
+    }
+}
+
+/// Implementation notes: is it possible to reduce indirection?
+/// Could we also skip the finalizer?
+///
+/// While putting T immediately inside the custom block as field(1)
+/// is tempting, GC would misalign it (UB) when moving.  Put a pointer to T instead.
+/// That optimisation would only work when alignment is the same as OCaml,
+/// meaning size_of<uintnat>.  It would also need to use different types.
```

Comment:
@g2p very interesting point. I was aiming for an implementation with `T` inside the block, but did not consider this.

Reply:
@zshipko I was checking the implementation in `ocaml-rs`, and values are stored "inline" in that case (just like in the code I am working on). Has that caused you any issue?

@g2p it seems to me that even if the OCaml GC moves the custom block, everything will still remain word-aligned. Can you elaborate a bit more on this issue? One thing I like about having the block own the value is that then I can mutably borrow the value, something that doesn't seem to be possible in this version.

Reply:
OCaml will align to words, but some Rust types require a larger alignment.
Another restriction is that this will only work for Unpin types.

Getting a mutable borrow will break strict aliasing for anything that has been through the OCaml heap, since it can be arbitrarily aliased on the OCaml side. Think `rust_func a a` called in OCaml.

Reply:
@g2p good points, thank you.

Reply:
> @zshipko I was checking the implementation in ocaml-rs, and values are stored "inline" in that case (just like in the code I am working on). Has that caused you any issue?

I haven't seen any issues personally, but it definitely sounds like a potential issue. One of my big blockers with integrating `ocaml-interop` is that the `ocaml-rs` custom value implementation relies on `Value` matching the OCaml C API - it would be nice to depend on this new version instead of trying to come up with a totally new implementation in `ocaml-rs`